hadoop-hdds/server-scm/src/test/java/org/apache/hadoop/hdds/scm/container/replication/health/TestEmptyContainerHandler.java

```diff
@@ -0,0 +1,194 @@
+        .setPendingOps(Collections.emptyList())
+        .setReport(new ReplicationManagerReport())
+        .setContainerInfo(containerInfo)
+        .setContainerReplicas(containerReplicas)
+        .build();
+
+    assertAndVerify(request, true, 5);
+  }
+
+  @Test
+  public void testEmptyAndClosedRatisContainerReturnsTrue() {
+    ContainerInfo containerInfo = ReplicationTestUtil.createContainerInfo(
+        ratisReplicationConfig, 1, CLOSED, 0L, 123L);
+    Set<ContainerReplica> containerReplicas = ReplicationTestUtil
+        .createReplicas(containerInfo.containerID(),
+            ContainerReplicaProto.State.CLOSED, 0L, 123L, 0, 0, 0);
+
+    ContainerCheckRequest request = new ContainerCheckRequest.Builder()
+        .setPendingOps(Collections.emptyList())
+        .setReport(new ReplicationManagerReport())
+        .setContainerInfo(containerInfo)
+        .setContainerReplicas(containerReplicas)
+        .build();
+
+    assertAndVerify(request, true, 3);
+  }
+
+  /**
+   * Handler should return false when key count is 0 but the container is not
+   * in CLOSED state.
+   */
+  @Test
+  public void testEmptyAndNonClosedECContainerReturnsFalse() {
+    ContainerInfo containerInfo = ReplicationTestUtil.createContainerInfo(
+        ecReplicationConfig, 1, CLOSING, 0L, 123L);
+
+    // though key count is 0, the container and its replicas are not CLOSED
+    Set<ContainerReplica> containerReplicas = ReplicationTestUtil
+        .createReplicas(containerInfo.containerID(),
+            ContainerReplicaProto.State.OPEN, 0L, 123L, 1, 2, 3, 4, 5);
+
+    ContainerCheckRequest request = new ContainerCheckRequest.Builder()
+        .setPendingOps(Collections.emptyList())
+        .setReport(new ReplicationManagerReport())
+        .setContainerInfo(containerInfo)
+        .setContainerReplicas(containerReplicas)
+        .build();
+
+    assertAndVerify(request, false, 0);
+  }
+
+  /**
+   * This test exists to verify that the definition of an empty container is
+   * 0 key count. Number of used bytes are not considered.
+   */
+  @Test
+  public void testNonEmptyRatisContainerReturnsFalse() {
+    ContainerInfo containerInfo = ReplicationTestUtil.createContainerInfo(
+        ratisReplicationConfig, 1, CLOSED, 5L, 123L);
+
+    // though container and its replicas are CLOSED, key count is not 0
+    Set<ContainerReplica> containerReplicas = ReplicationTestUtil
+        .createReplicas(containerInfo.containerID(),
+            ContainerReplicaProto.State.CLOSED, 5L, 123L, 0, 0, 0);
+
+    ContainerCheckRequest request = new ContainerCheckRequest.Builder()
+        .setPendingOps(Collections.emptyList())
+        .setReport(new ReplicationManagerReport())
+        .setContainerInfo(containerInfo)
+        .setContainerReplicas(containerReplicas)
+        .build();
+
+    assertAndVerify(request, false, 0);
+  }
+
+  @Test
+  public void testEmptyContainerWithNonZeroBytesUsedReturnsTrue() {
```

Comment:
Is this test bascically the same as `testEmptyAndClosedRatisContainerReturnsTrue()`?

Reply:
Yes, their content is the same. While `testEmptyAndClosedRatisContainerReturnsTrue` implicitly checks for 0 key count and > 0 bytes used, I added this test to explicitly assert the correct behaviour. My reasoning is that `testEmptyAndClosedRatisContainerReturnsTrue` could change later and we might lose that implicit check.

Reply:
Basically trying to replicate this test from `TestLegacyReplicationManager` where key count is 0 but bytes used isn't:
```
  /**
   * A closed empty container with all the replicas also closed and empty
   * should be deleted.
   * A container/ replica should be deemed empty when it has 0 keyCount even
   * if the usedBytes is not 0 (usedBytes should not be used to determine if
   * the container or replica is empty).
   */
  @Test
  public void testDeleteEmptyContainer() throws Exception {
    runTestDeleteEmptyContainer(3);
  }

  Void runTestDeleteEmptyContainer(int expectedDelete) throws Exception {
    // Create container with usedBytes = 1000 and keyCount = 0
    final ContainerInfo container = createContainer(LifeCycleState.CLOSED, 1000,
        0);
    addReplica(container, new NodeStatus(IN_SERVICE, HEALTHY), CLOSED);
    addReplica(container, new NodeStatus(IN_SERVICE, HEALTHY), CLOSED);
    // Create a replica with usedBytes != 0 and keyCount = 0
    addReplica(container, new NodeStatus(IN_SERVICE, HEALTHY), CLOSED, 100, 0);

    assertDeleteScheduled(expectedDelete);
    return null;
  }
```
We can delete it if you think it's unnecessary